shared/constants/chat2/meta.js

```diff
@@ -0,0 +1,325 @@
+    conversationIDKey: Types.stringToConversationIDKey(i.convID),
+    inboxVersion: i.version,
+    isMuted: i.status === RPCChatTypes.commonConversationStatus.muted,
+    membershipType: conversationMemberStatusToMembershipType(i.memberStatus),
+    participants,
+    resetParticipants,
+    snippet: i.localMetadata ? i.localMetadata.snippet : '',
+    supersededBy: supersededBy ? Types.stringToConversationIDKey(supersededBy) : null,
+    supersededByCausedBy,
+    supersedes: supersedes ? Types.stringToConversationIDKey(supersedes) : null,
+    supersedesCausedBy,
+    teamType: getTeamType(i),
+    teamname,
+    timestamp: i.time,
+    tlfname: i.name,
+    trustedState: 'untrusted',
+  })
+}
+
+const conversationMetadataToMetaSupersedeInfo = (metas: ?Array<RPCChatTypes.ConversationMetadata>) => {
+  const meta: ?RPCChatTypes.ConversationMetadata = (metas || []).find(
+    m => m.idTriple.topicType === RPCChatTypes.commonTopicType.chat && m.finalizeInfo
+  )
+
+  return {
+    conversationIDKey: meta ? supersededConversationIDToKey(meta.conversationID) : null,
+    username: meta && meta.finalizeInfo ? meta.finalizeInfo.resetUser : null,
+  }
+}
+
+const getTeamType = ({teamType, membersType}) => {
+  if (teamType === RPCChatTypes.commonTeamType.complex) {
+    return 'big'
+  } else if (membersType === RPCChatTypes.commonConversationMembersType.team) {
+    return 'small'
+  } else {
+    return 'adhoc'
+  }
+}
+
+// Upgrade a meta, try and keep exising values if possible to reduce render thrashing in components
+// Enforce the verions only increase and we only go from untrusted to trusted, etc
+export const updateMeta = (
+  old: Types.ConversationMeta,
+  meta: Types.ConversationMeta
+): Types.ConversationMeta => {
+  // Older/same version and same state?
+  if (meta.inboxVersion <= old.inboxVersion && meta.trustedState === old.trustedState) {
+    return old
+  }
+
+  const participants = old.participants.equals(meta.participants) ? old.participants : meta.participants
+  const rekeyers = old.rekeyers.equals(meta.rekeyers) ? old.rekeyers : meta.rekeyers
+  const resetParticipants = old.resetParticipants.equals(meta.resetParticipants)
+    ? old.resetParticipants
+    : meta.resetParticipants
+
+  return meta.withMutations(m => {
+    m.set('channelname', meta.channelname || old.channelname)
+    m.set('paginationKey', old.paginationKey)
+    m.set('paginationMoreToLoad', old.paginationMoreToLoad)
+    m.set('orangeLineOrdinal', old.orangeLineOrdinal)
+    m.set('participants', participants)
+    m.set('rekeyers', rekeyers)
+    m.set('resetParticipants', resetParticipants)
+    m.set('teamname', meta.teamname || old.teamname)
+  })
+}
+
+const parseNotificationSettings = (notifications: ?RPCChatTypes.ConversationNotificationInfo) => {
+  let notificationsDesktop = 'never'
+  let notificationsGlobalIgnoreMentions = false
+  let notificationsMobile = 'never'
+
+  // Map this weird structure from the daemon to something we want
```

Comment:
I also don't understand this. Why not just add some function to pull this info out of the type we give you?

Reply:
thats what this function is? i don't understand